### DISCUSSION TITLE
Hide "private" properties with an underscore from users by default.

### DISCUSSION THREAD

**Discussion by willnationsdev** [godot-extended-libraries/godot-ideas#33] [Ideas]
Writing this quick idea down so that I can write a more detailed proposal later.

## Context
Using properties in scripts from different object types.

## Problem
Users can often times directly see and access "private" properties and data in other types.

## Suggestion
Filter properties with `PROPERTY_USAGE_SCRIPT_VARIABLE` and a leading underscore in their name OUT of the `get_property_list()` method's results. This will prevent autocompletion from functioning properly. Need to figure out how to make the basic idea work in more contexts though, e.g. the Inspector may still want to expose a property with a leading underscore and we need to make that painless.

## Solution
Not sure yet.

**Comment by ghost** [comment]
shouldn't this be an option? wouldn't this lead to inconsistent behaviour \ unintuitive behaviour?

private, as in, c++ like, classes private? i don't really understand, sorry
i actually have no idea what this whole thing is supposed to do

wait what, why do these have the same name, that seems kinda terrible
![](https://i.imgur.com/JNHNPUi.png)
![](https://i.imgur.com/z6pS2Uv.png)

the documentation is kinda vague and i have no idea what this supposed to do, i have no idea what counts as a "property", what, like ints, strings? some kind of math? a flag? what is it, it could be anything, but there's barely any information

**Comment by willnationsdev** [comment]
> shouldn't this be an option?

Yeah, I could see this being in the ProjectSettings or something.

> wouldn't this lead to inconsistent behaviour \ unintuitive behaviour?

Perhaps. But there is precedent in other languages though, for example, Dart.

> private, as in, c++ like, classes private? i don't really understand, sorry

Yes, as in, having private properties and methods that are accessible within the class, but which are not visible to logic outside of the class.

> i actually have no idea what this whole thing is supposed to do

If have `MyClass` with a 3 public properties and a dozen private properties, I would expect the autocompletion dropdown in the ScriptEditor to only display the 3 public properties and not 15 properties when I do an `my_variable.<wait_for_autocompletion>`.

GDScript does not yet have a way of indicating whether something is private or public. I could imagine a scenario where underscore-prefixed methods, properties, and constants are inaccessible outside of the class. Although, it definitely gets tricky since signal callbacks are regularly underscore-prefixed as well.

A full `private` keyword might be best.

> wait what, why do these have the same name, that seems kinda terrible

The first is a virtual engine callback. You return an Array of Dictionaries with specific key-value pairs and it will build corresponding PropertyInfo structs in the C++ engine to be appended to the property list for the class. This enables users to procedurally generate properties based on logical processes rather than declaring them statically.

The second is the *actual* Object method to request and get back the full list of properties (again, as an Array of Dictionaries). With it, you can get a list of all properties defined by an Object instance. This will include all PropertyInfo Dictionaries from the inherited classes, the underlying Object, AND the Script + its `_get_property_list()` callback results, if defined.

> the documentation is kinda vague and i have no idea what this supposed to do, i have no idea what counts as a "property", what, like ints, strings? some kind of math? a flag? what is it, it could be anything, but there's barely any information

These methods have to do with the way Objects in Godot track and define their reflection data. The engine defines `PropertyInfo` and `MethodInfo` structs internally. Anyone of these struct instances can be represented as a Dictionary. Properties and constants are described using a PropertyInfo; methods and signals are described using a MethodInfo.

A PropertyInfo contains

1. the data `type` as a `TYPE_*` enum value.
1. the property `name` as a string.
1. a `hint` enum value that indicates any special way in which the data should be interpreted.
1. a `hint_string` string that supplements the `hint` enum with more information
    - Example: the `PROPERTY_HINT_ENUM` is used with `TYPE_INT` properties to indicate that the integer is an enum. This tells the Editor's Inspector to render the integer differently, i.e. as a dropdown with options. The `hint_string` field presents a comma-delimited list of the text that should be used for each option, e.g. `"Aces,Hearts,Spades,Clubs"`
1. a `class_name` string which specifies the name of the class type in cases where the data type is `TYPE_OBJECT`, e.g. it might be `"Node"`.

**Comment by ghost** [comment]
@RaTcHeT302

I have a passing familiarity with Python syntax, so seeing dunders in code isn't too shocking, personally. But I imagine someone trying to learn GDScript as a first language from the docs would be wondering why `SomeNames` look `_funny`.

The [GDScript style guide](https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_styleguide.html#naming-conventions) explains the use of underscores, but it's hard to find if you don't know what you're looking for:

*Prepend a single underscore (_) to virtual methods functions the user must override, private functions, and private variables:*
```
var _counter = 0
func _recalculate_path():
```

Learning first time scripting from the docs is sketchy. Good documentation should be terse and accurate, if not a bit dry. You're much better off using books + videos + forums that are more hand-holdy.

@willnationsdev

GDScript is getting a [refactor](https://godotengine.org/article/gdscript-progress-report-new-gdscript-now-merged). Maybe it'll open up some possibilities.

**Comment by willnationsdev** [comment]
@frogonics Yes, with the GDScript rewrite, we are getting preliminary annotation support. With that, it will likely be possible to add a `@private` annotation to the language for properties and methods to indicate that they should not be accessible from the traditional get/set/get\_property\_list/call methods. Actually wouldn't be that complicated of a feature to implement, if my reading of the new GDScript parser is correct.
